src/rail_gui.cpp

```diff
@@ -956,37 +1004,131 @@ struct BuildRailStationWindow : public PickerWindowBase {
-				count++;
-			}
-			this->vscroll->SetCount(count);
-			this->vscroll->SetPosition(Clamp(_railstation.station_class - 2, 0, std::max(this->vscroll->GetCount() - this->vscroll->GetCapacity(), 0)));
-
 			NWidgetMatrix *matrix = this->GetWidget<NWidgetMatrix>(WID_BRAS_MATRIX);
 			matrix->SetScrollbar(this->vscroll2);
 			matrix->SetCount(_railstation.station_count);
 			matrix->SetClicked(_railstation.station_type);
+
+			EnsureSelectedStationClassIsVisible();
+
+			this->SetFocusedWidget(WID_BRAS_FILTER_EDITBOX);
 		}
+
+		this->InvalidateData();
 	}
 
 	virtual ~BuildRailStationWindow()
 	{
 		DeleteWindowById(WC_SELECT_STATION, 0);
 	}
 
+	/** Sort station classes by StationClassID. */
+	static bool StationClassIDSorter(StationClassID const &a, StationClassID const &b)
+	{
+		return a < b;
+	}
+
+	/** Filter station classes by class name. */
+	static bool CDECL TagNameFilter(StationClassID const * sc, StringFilter &filter)
+	{
+		char buffer[DRAW_STRING_BUFFER];
+		GetString(buffer, StationClass::Get(*sc)->name, lastof(buffer));
+
+		filter.ResetState();
+		filter.AddLine(buffer);
+		return filter.GetState();
+	}
+
+	/** Builds the filter list of available station classes. */
+	void BuildStationClassesAvailable()
+	{
+		if (!this->station_classes.NeedRebuild()) return;
+
+		this->station_classes.clear();
+
+		for (uint i = 0; i < StationClass::GetClassCount(); i++) {
+			StationClassID station_class_id = (StationClassID)i;
+			if (station_class_id == StationClassID::STAT_CLASS_WAYP) {
+				// Skip waypoints.
+				continue;
+			}
+			StationClass *station_class = StationClass::Get(station_class_id);
+			if (station_class->GetUISpecCount() == 0) continue;
+			station_classes.push_back(station_class_id);
+		}
+
+		if (_railstation.newstations) {
+			this->station_classes.Filter(this->string_filter);
+			this->station_classes.shrink_to_fit();
+			this->station_classes.RebuildDone();
+			this->station_classes.Sort();
+
+			this->vscroll->SetCount((uint)this->station_classes.size());
+		}
+	}
+
+	/**
+	 * Checks if the previously selected current station class and station
+	 * can be shown as selected to the user when the dialog is opened.
+	 */
+	void SelectClassAndStation()
+	{
+		if (_railstation.station_class == StationClassID::STAT_CLASS_DFLT) {
+			/* This happens during the first time the window is open during the game life cycle. */
+			this->SelectOtherClass(StationClassID::STAT_CLASS_DFLT);
+		} else {
+			/* Check if the previously selected station class is not available anymore as a
+			 * result of starting a new game without the corresponding NewGRF. */
+			bool available = false;
+			for (uint i = 0; i < StationClass::GetClassCount(); ++i) {
+				if ((StationClassID)i == _railstation.station_class) {
+					available = true;
+					break;
+				}
+			}
+
+			if (available) {
+				this->SelectOtherClass(_railstation.station_class);
+			} else {
+				this->SelectOtherClass(this->station_classes[0]);
```

Comment:
I assume there's no possibility of station_classes being empty? Maybe a comment/assert to "verify" it :)

Reply:
Changing to StationClassID::STAT_CLASS_DFLT if not available as it's always the 1st in the list. Thanks!